### PR TITLE
Run cppcheck on PRs

### PR DESCRIPTION
**Pull request by Flakebi**
Add a CI run with cppcheck, a static C++ code analyzer.

Only the ‘Run cppcheck on PRs’ commit is part of the pr, the other one is from #712.

### CODE REVIEW

**Comment by kuhar** [comment]
Since we have working clang builds, I think clang tidy and/or clang static analyzer should also be easy to wire in.
The base amdvlk builds produce compilation database with similar use cases in mind :smiley:

File: .github/workflows/check-cppcheck-llpc.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      - name: Setup environment
+        run: |
+          sudo apt-get install -yqq cppcheck
+      - name: Checkout LLPC
+        run: |
+          git clone https://github.com/${GITHUB_REPOSITORY}.git .
+          git fetch origin +${GITHUB_SHA}:${GITHUB_REF} --update-head-ok
+          git checkout ${GITHUB_SHA}
+      - name: Run cppcheck
+        run: cppcheck -q -j8 --error-exitcode=1 .
```
[the remark this answers is not in the document]
Comment (Flakebi) [reply .github/workflows/check-cppcheck-llpc.yml:20]: cppcheck seems to be io-limited, so adding more threads makes it run faster, even if there are more than the number of cores. Testing locally, I even got an improvement with `-j128` although it was only minor (unfortunately I don’t have that many cores).
I thought 8 is a great number, there was not much improvement after that (on my pc), though jumping from 4 to 8 cut the time in half.

**Comment by Flakebi** [comment]
> How effective can it be without having access to the rest of the codebase (e.g., PAL, XGL, etc.) and without having a compilation database with the compilation flags.

I don’t know much about cppcheck, judging from the examples it seems like a common think to run cppcheck on single files or folders, so not having the whole source does not seem to be a problem.

> How do you want to deal with false positives? So far all CI checks, except warnings, only complained on hard failures.

As far as I understand it, there are no false positives with the default options. `--inconclusive` and `--enable=warning` add warnings but they look mostly unnecessary. #714 fixes one of the more useful warnings but even that one is not really important.

> Since we have working clang builds, I think clang tidy and/or clang static analyzer should also be easy to wire in.
> The base amdvlk builds produce compilation database with similar use cases in mind :smiley:

Nice!

File: .github/workflows/check-cppcheck-llpc.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kuhar) [reply .github/workflows/check-cppcheck-llpc.yml:20]: Would it make sense to use `nproc * 4` then?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 45ff3b7bc20499e457c6adc2ad3146e03b6c33f6
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev 704f4adddb6867d633431c2c9c2c0bfa6d294ce2</li>
<li>XGL: dev 9c4fe791366e730131a0d99a7519855068f216ef</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 3a1103f746f4d782271bd32d67274d6260c5dc7f</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
Cppcheck can consume compilation database http://cppcheck.sourceforge.net/manual.pdf

```
cppcheck --project=compile_commands.json
```

and exclude specified directories:

```
cppcheck -isrc/b -isrc/c
```

I think it would make more sense give it the whole base image and tell it to check llpc only. Later, we can extend the same workflow to run other linters.

> As far as I understand it, there are no false positives with the default options
Modulo implementation bugs and other things we don't know about yet. But I think it'll be perfectly fine to enable it and revisit this when it actually becomes an issue.

File: .github/workflows/check-cppcheck-llpc.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Flakebi) [reply .github/workflows/check-cppcheck-llpc.yml]: Good idea

**Comment by Flakebi** [comment]
I tried `--project=compile_commands.json` which showed some errors (few in LLVM, some in other driver parts, more in SPIRV-tools) but I didn’t manage to ignore them, neither `-illvm`, nor `../../llvm-project/llvm` nor an absolute path worked.
One thing I noticed is that it does not detect errors anymore which are in parts of the source which is not compiled.

Do we gain something by handing it the compile commands?
Giving cppcheck access to preprocessor definitions seems to make it worse (i.e. less errors detected).
The other thing that it can learn are include paths. As we don’t want to check header files outside llpc, we would need to exclude those. The manual actually recommends to *not include* the standard library headers in include paths because it slows down checking.

> I think it would make more sense give it the whole base image and tell it to check llpc only.

Which one of the base images would make sense?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 507f6d3668add24c30683fde8b3d44c237eecc26
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 85170384d0ee5447ac267d478f7a21bc036ea71c</li>
<li>SPVGEN: dev 704f4adddb6867d633431c2c9c2c0bfa6d294ce2</li>
<li>XGL: dev 9c4fe791366e730131a0d99a7519855068f216ef</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 3a1103f746f4d782271bd32d67274d6260c5dc7f</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
> I tried `--project=compile_commands.json` which showed some errors (few in LLVM, some in other driver parts, more in SPIRV-tools) but I didn’t manage to ignore them, neither `-illvm`, nor `../../llvm-project/llvm` nor an absolute path worked.
> One thing I noticed is that it does not detect errors anymore which are in parts of the source which is not compiled.

So it seems like it just runs on all files listed in compile_commands.
Perhaps we can just reduce the compile_commands file and filter out the non-llpc paths from there? This should be a simple `jq` query.
Re checking not compiled code, I think this makes perfect sense, e.g., there's a lot of contradictory code in linux and windows ifdefs.

> Do we gain something by handing it the compile commands?
> Giving cppcheck access to preprocessor definitions seems to make it worse (i.e. less errors detected).
> The other thing that it can learn are include paths. As we don’t want to check header files outside llpc, we would need to exclude those. The manual actually recommends to _not include_ the standard library headers in include paths because it slows down checking.

You cannot check the code you don't know about. Two extremes here would (1) be having a whole-program information (think full-LTO) and doing global analysis, and (2) looking at a function/code block at a time and doing very simple syntactic checks without understanding what the code means.
If you are in (1) you can chose to drop some information that is not useful/desired (e.g., headers, function definitions, etc.) to gain speed or avoid internal bugs.

From what you are describing here it seems to me that cppcheck struggles with parsing / analyzing the LLPC code (no big surprise here, LLVM headers are complicated) and only works on the most basic level like (2). The kind of bugs you can catch like that (without AST) is very limited, especially if you try hard to avoid false positives but do we mind? I'm not familiar with cppcheck and don't know if it produces helpful findings when configured like this.

Perhaps it would make sense to run cppcheck in the dummy mode and in addition run clang-tidy with compilation database enabled?
 
> > I think it would make more sense give it the whole base image and tell it to check llpc only.
> 
> Which one of the base images would make sense?
For clang-based tools I'd pick the vanilla clang image, or the image with shader cache. I don't think it matters much.

In general, I think that more CI checks would be great, but at the same time I'm skeptical that a linter that doesn't know how to even build/parse code will find anything useful. But if you can argue otherwise, or have good experience with this sort of setup, I think we should give it a try and see how it performs. Adding a new CI config is cheap and easy, and so is removing an existing one.

**Comment by Flakebi** [comment]
> Perhaps it would make sense to run cppcheck in the dummy mode and in addition run clang-tidy with compilation database enabled?

That sounds good. As you said, cppcheck’s analysis is not deep. It found two bugs, which is nice (#712).

**Comment by kuhar** [comment]
> > Perhaps it would make sense to run cppcheck in the dummy mode and in addition run clang-tidy with compilation database enabled?
> 
> That sounds good. As you said, cppcheck’s analysis is not deep. It found two bugs, which is nice (#712).

If you are happy with this mode, then it sound OK for me.

I saw cppcheck [implements value flow analysis](http://cppcheck.sourceforge.net/devinfo/doxyoutput/valueflow_8cpp.html) and can use [Z3 for constraint solving](http://cppcheck.sourceforge.net/devinfo/doxyoutput/namespaceExprEngine.html), so I think it would be a good idea to first figure out if we can run it properly to get the best out of it, though.

**Comment by kuhar** [comment]
One other thing you could do to make it more reasonable would be to take all the preprocessor defines from compile_commands.json for some file (e.g., from amdllpc.cpp) and give it to cppcheck in the 'dummy mode'. Something like this: https://github.com/alicevision/geogram/blob/master/tools/cppcheck.sh

**Comment by Flakebi** [comment]
The value flow analysis sounds really interesting. Like `--clang`, it is only available in version 2, which was published two weeks ago and is not in the Ubuntu repositories, so we would need to compile it ourselves.
I’ll definitely try this.

Judging from the output while running cppcheck without `--quit`, it checks files with combinations of preprocessor definitions if it encounters them in a file (that’s how it discovered a syntax error in a configuration that nobody compiled in the last months…). I’ll poke around tomorrow if it reports something when adding preprocessor defines manually.

**Comment by kuhar** [comment]
You can install recent versions with snap: https://snapcraft.io/cppcheck.

**Comment by JacobHeAMD** [comment]
There are some usages possibly is against cppcheck, for example,
114 class Device
 115 {
...
  722     // This goes last.  The memory for the rest of the array is calculated dynamically based on the number of GPUs in
 723     // use.
 **724     PerGpuInfo              m_perGpu[1];**
725 };

cppcheck will report "[api/include/vk_device.h:533]: (error) Array 'm_perGpu[1]' accessed at index 1, which is out of bounds." for this case. And there are same usage in PAL. Not sure if is it fine for LLPC?

**Comment by Flakebi** [comment]
Interesting point.
It seems like there are no such constructs in LLPC, at least cppcheck does not complain about anything.
While technically this construct works on all compilers (and probably will continue to work), cppcheck still has a point that it is undefined behavior.
As written by the C committee ([stackoverflow thread with more context](https://stackoverflow.com/a/12412534)):
>The validity of this construct has always been questionable. In the response to one Defect Report, the Committee decided that it was undefined behavior because the array p->items contains only one item, irrespective of whether the space exists.

**Comment by kuhar** [comment]
IIRC this example falls under potential strict aliasing violations (e.g., path TBAA in clang/llvm). A tool aware of `-fno-strict-aliasing` in build options should not complain.

**Comment by kuhar** [comment]
cppcheck and clang-tidy/CSA allow disabling individual checks, so I don't thing it's a big issue if it complains about this pattern

**Comment by Flakebi** [comment]
I tried the version with Z3, it didn’t find anything new for llpc.
It seems like running snap (to get an up-to-date cppcheck) in a docker container is difficult to do, at least I didn’t get it to work.

**Comment by kuhar** [comment]
I installed it with snap with no issues: https://github.com/kuhar/llpc/pull/2.
It even found one more macro issue :stuck_out_tongue:

**Comment by Flakebi** [comment]
Ah, I tried to do that in docker to run with `--project=compile_commands.json`.

> It even found one more macro issue

Nice!

**Comment by kuhar** [comment]
You can extract the build tree from the docker image, copy it over to the github runner, and use the cppcheck from the github workflow on it fairly easily: https://github.com/kuhar/llpc/pull/2/files

**Comment by kuhar** [comment]
> Edit: Looks like a false positive?
Yep. That's why we'd rather use tools that can actually parse the sources

**Comment by kuhar** [comment]
Hi Sebastian,

Just wanted to followup on one thing from the past CI PR. Do you have some progress on making the assertions CI builds actually run with assertions on? This will be very useful to have.

**Comment by Flakebi** [comment]
I tried to get it working with the snap (the build needs to be moved to e.g. the home directory because others are not mounted in the snap container) but it showed no issues. When running locally, it complains about a few things in pal and llvm, so it didn’t work on GitHub.

As the version with Z3 and with `--project` does not find anything interesting for llpc at the moment (running locally), are you ok with running the check as it is in this pr?
(Also considering that running it on the docker container code takes 4 min compared to ½ min without)

**Comment by kuhar** [comment]
OK, fine for me. Thanks for checking that.

**Pull request merged**